// ts-type-guards/dist/array.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const is_1 = require("./is");
function isArrayOfBooleans(x) {
    return isArrayOfLike(true)(x);
}
exports.isArrayOfBooleans = isArrayOfBooleans;
function isArrayOfNumbers(x) {
    return isArrayOfLike(1)(x);
}
exports.isArrayOfNumbers = isArrayOfNumbers;
function isArrayOfStrings(x) {
    return isArrayOfLike("")(x);
}
exports.isArrayOfStrings = isArrayOfStrings;
function isArrayOfSymbols(x) {
    return isArrayOfLike(Symbol())(x);
}
exports.isArrayOfSymbols = isArrayOfSymbols;
function isArrayOfNulls(x) {
    return isArrayOfLike(null)(x);
}
exports.isArrayOfNulls = isArrayOfNulls;
function isArrayOfUndefineds(x) {
    return isArrayOfLike(undefined)(x);
}
exports.isArrayOfUndefineds = isArrayOfUndefineds;
function isArrayOfPrimitives(x) {
    return is_1.is(Array)(x) && x.every(is_1.isPrimitive);
}
exports.isArrayOfPrimitives = isArrayOfPrimitives;
function isArrayOfObjects(x) {
    return is_1.is(Array)(x) && x.every(is_1.isNonPrimitive);
}
exports.isArrayOfObjects = isArrayOfObjects;
function isArrayOf(type) {
    return (xs) => is_1.is(Array)(xs) && xs.every(is_1.is(type));
}
exports.isArrayOf = isArrayOf;
function isArrayOfLike(reference) {
    return (x) => is_1.is(Array)(x) && x.every(is_1.isLike(reference));
}
exports.isArrayOfLike = isArrayOfLike;
