// ts-type-guards/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var is_1 = require("./is");
exports.isBoolean = is_1.isBoolean;
exports.isBooleanLike = is_1.isBooleanLike;
exports.isNumber = is_1.isNumber;
exports.isNumberLike = is_1.isNumberLike;
exports.isString = is_1.isString;
exports.isStringLike = is_1.isStringLike;
exports.isSymbol = is_1.isSymbol;
exports.isNull = is_1.isNull;
exports.isUndefined = is_1.isUndefined;
exports.isSomething = is_1.isSomething;
exports.isNothing = is_1.isNothing;
exports.isPrimitive = is_1.isPrimitive;
exports.isNonPrimitive = is_1.isNonPrimitive;
exports.is = is_1.is;
exports.isLike = is_1.isLike;
var array_1 = require("./array");
exports.isArrayOfBooleans = array_1.isArrayOfBooleans;
exports.isArrayOfNumbers = array_1.isArrayOfNumbers;
exports.isArrayOfStrings = array_1.isArrayOfStrings;
exports.isArrayOfSymbols = array_1.isArrayOfSymbols;
exports.isArrayOfNulls = array_1.isArrayOfNulls;
exports.isArrayOfUndefineds = array_1.isArrayOfUndefineds;
exports.isArrayOfPrimitives = array_1.isArrayOfPrimitives;
exports.isArrayOfObjects = array_1.isArrayOfObjects;
exports.isArrayOf = array_1.isArrayOf;
exports.isArrayOfLike = array_1.isArrayOfLike;
var only_1 = require("./only");
exports.onlyBooleans = only_1.onlyBooleans;
exports.onlyNumbers = only_1.onlyNumbers;
exports.onlyStrings = only_1.onlyStrings;
exports.onlySymbols = only_1.onlySymbols;
exports.onlyNulls = only_1.onlyNulls;
exports.onlyUndefineds = only_1.onlyUndefineds;
exports.onlyPrimitives = only_1.onlyPrimitives;
exports.onlyObjects = only_1.onlyObjects;
exports.only = only_1.only;
exports.onlyLike = only_1.onlyLike;
