// ts-type-guards/dist/only.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const is_1 = require("./is");
function onlyBooleans(xs) {
    return xs.filter(is_1.isBoolean);
}
exports.onlyBooleans = onlyBooleans;
function onlyNumbers(xs) {
    return xs.filter(is_1.isNumber);
}
exports.onlyNumbers = onlyNumbers;
function onlyStrings(xs) {
    return xs.filter(is_1.isString);
}
exports.onlyStrings = onlyStrings;
function onlySymbols(xs) {
    return xs.filter(is_1.isSymbol);
}
exports.onlySymbols = onlySymbols;
function onlyNulls(xs) {
    return xs.filter(is_1.isNull);
}
exports.onlyNulls = onlyNulls;
function onlyUndefineds(xs) {
    return xs.filter(is_1.isUndefined);
}
exports.onlyUndefineds = onlyUndefineds;
function onlyPrimitives(xs) {
    return xs.filter(is_1.isPrimitive);
}
exports.onlyPrimitives = onlyPrimitives;
function onlyObjects(xs) {
    return xs.filter(is_1.isNonPrimitive);
}
exports.onlyObjects = onlyObjects;
function only(type) {
    return (xs) => xs.filter(is_1.is(type));
}
exports.only = only;
function onlyLike(reference) {
    return (xs) => xs.filter(is_1.isLike(reference));
}
exports.onlyLike = onlyLike;
