// squiss-ts/dist/s3Utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deleteBlob = exports.getBlob = exports.uploadBlob = exports.S3_MARKER = void 0;
const uuid_1 = require("uuid");
const messageSizeUtils_1 = require("./messageSizeUtils");
exports.S3_MARKER = '__SQS_S3__';
const uploadBlob = (s3, bucket, blob, prefix) => {
    const key = `${prefix}${(0, uuid_1.v4)()}`;
    const size = (0, messageSizeUtils_1.getSizeInBytes)(blob);
    return s3.putObject({
        Body: blob,
        Bucket: bucket,
        Key: key,
        ContentLength: size,
    })
        .then(() => {
        return Promise.resolve({
            uploadSize: size,
            bucket,
            key,
        });
    });
};
exports.uploadBlob = uploadBlob;
const getBlob = async (s3, uploadData) => {
    const key = uploadData.key;
    const bucket = uploadData.bucket;
    const s3Result = await s3.getObject({
        Key: key,
        Bucket: bucket,
    });
    const s3ResponseStream = s3Result.Body;
    const chunks = [];
    for await (const chunk of s3ResponseStream) {
        chunks.push(chunk);
    }
    return Buffer.concat(chunks).toString('utf-8');
};
exports.getBlob = getBlob;
const deleteBlob = (s3, uploadData) => {
    const key = uploadData.key;
    const bucket = uploadData.bucket;
    return s3.deleteObject({
        Key: key,
        Bucket: bucket,
    })
        .then(() => {
        return Promise.resolve();
    });
};
exports.deleteBlob = deleteBlob;
